Twitter processor

brew opened the pull request:
This pull request fixes #17 and is connected to #3.

* [x] I've added tests to cover the proposed changes

Changes proposed in this pull request:

- Fleshes out the socialmedia pipeline
- Adds a twitter processor
  - adds `followers`, `mentions` and `interactions`
- Adds an `SQLDatastore` interface

Processor retrieves two types of metrics, 'lifetime' metrics, such as `followers`, that are retrieved from the api, and used as-is without processing, and 'cumulative' metrics, such as `mentions` and `interactions` that are retrieved for a specific time-frame, then added to the latest stored result.

akariv commented on datapackage_pipelines_measure/pipeline_steps/social_media.py:
```diff
@@ -1,23 +1,74 @@
+    steps.append(('concatenate', {
+        'target': {
+            'name': 'social-media',
+            'path': 'data/social-media.json'},
```
Why do you add a path here? Why the json extension?

brew replied:
The path is useful if I want to `dump.to_path` in development. I have a patch lined up that will dump the latest to path, only there's a DEVELOPMENT flag in the env.

The `.json` extension is a mistake, and interestingly, is ignored when the resource file is dumped, and a `.csv` extension is used instead.

akariv commented on datapackage_pipelines_measure/processors/add_twitter_resource.py:
```diff
@@ -0,0 +1,285 @@
+safe_entity = _get_safe_entity(entity)
+resource = {
+    'name': safe_entity,
+    'path': 'data/{}.json'.format(safe_entity)
```
Why the json?

brew replied:
Again, this should be a `.csv`. Though, most of the time this path is never used, as the resources are concatenated later in the pipeline.

akariv commented on datapackage_pipelines_measure/processors/add_twitter_resource.py:
```diff
@@ -0,0 +1,285 @@
+datapackage['resources'].append(resource)
+
+# Make a single-item generator from the resource_content
+resource_content = itertools.chain([resource_content])
```
that's not really needed.

brew replied:
If I don't do this, I get an error from `concatenate`:

```
INFO    :Main                            :concatenate: /Users/brew/virtualenvs/measure/lib/python3.6/site-packages/jsontableschema/model.py:48: UserWarning: Class models.SchemaModel is deprecated [v0.7-v1)
INFO    :Main                            :concatenate:   warnings.warn(message, UserWarning)
INFO    :Main                            :concatenate: Traceback (most recent call last):
INFO    :Main                            :concatenate:   File "/Users/brew/virtualenvs/measure/src/datapackage-pipelines/datapackage_pipelines/specs/../lib/concatenate.py", line 119, in <module>
INFO    :Main                            :concatenate:     spew(datapackage, new_resource_iterator(resource_iterator))
INFO    :Main                            :concatenate:   File "/Users/brew/virtualenvs/measure/src/datapackage-pipelines/datapackage_pipelines/wrapper/wrapper.py", line 59, in spew
INFO    :Main                            :concatenate:     for rec in res:
INFO    :Main                            :concatenate:   File "/Users/brew/virtualenvs/measure/src/datapackage-pipelines/datapackage_pipelines/specs/../lib/concatenate.py", line 99, in concatenator
INFO    :Main                            :concatenate:     in row.items()
INFO    :Main                            :concatenate: AttributeError: 'str' object has no attribute 'items'
```

akariv commented on datapackage_pipelines_measure/datastore/sql.py:
```diff
@@ -0,0 +1,42 @@
+from sqlalchemy.ext.automap import automap_base
```
See later comment on why I think you don't need this.

akariv commented on datapackage_pipelines_measure/processors/add_twitter_resource.py:
```diff
@@ -0,0 +1,285 @@
+    'project_id': project_id
+}
+datastore = get_datastore()
+entity_last_run = datastore.get_latest_from_table(filter_object,
```
Not sure where this comment belongs so I'm putting it here.

I think that a simpler solution would be to always query 3 times for the past three days (i.e. today, yesterday & the day before that).
Then you generate 3 separate records for each of these query results.

When writing to the DB, you use the timestamp (and probably entity, entity_type and source) as key for updating the DB.

This way you don't need to store any permanent state (which will be deleted once the container is re-deployed).
Also, I think that conceptually aggregation can (and should) be done separately, and not in the raw collector - either in a separate pipeline or in a simple SQL query.

brew replied:
After some discussion offline, we decided not to aggregate data in the processor, instead leaving aggregation as a task for the reporting layer (Redash/SQL). This should simplify collection.

akariv commented on datapackage_pipelines_measure/processors/add_twitter_resource.py:
```diff
@@ -0,0 +1,285 @@
+        end_date.strftime(TWITTER_API_DATE_RANGE_FORMAT)
+    )
+
+resource_content = {
```
no timestamp in record? 
I think that we should collect the raw data here and do aggregations later.

brew replied:
Timestamp is added in a later processor in the pipeline.

brew commented:
@akariv I've refactored the twitter processor so it no longer aggregates locally. 

I decided to keep the timestamp as it is, as it represents the datetime the data was collected, and I'd like to differentiate between the time collected (metadata) and the date for the data we're interested (which itself is data). So I've added a `date` column to represent the data date.
